bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs/diff/osgi/Commands.scala

```diff
@@ -0,0 +1,160 @@
+          CmdArgs(locationFor(id), id)
+        }
+        case 0 => lastProgId.toRightDisjunction(usage(cmd)).map { id =>
+          CmdArgs(locationFor(id), id)
+        }
+        case _ => usage(cmd).left
+      }
+
+    def remember(args: CmdArgs) {
+      lastProgId = Some(args.id)
+      reg.register(args.id, args.loc)
+    }
+
+    type VcsOp = (SPProgramID, Peer) => TryVcs[String]
+
+    def doVcsOp(cmd: String, args: Array[String])(op: VcsOp): String =
+      (for {
+        ca  <- parseArgs(cmd, args)
+        res <- op(ca.id, ca.loc).leftMap { f =>
+                 VcsFailure.explain(f, ca.id, cmd, Some(ca.loc))
+               }
+      } yield {
+        remember(ca)
+        res
+      }).toEither.merge
+
+    def timed[A](body: => A): (A, Long) = {
+      val start = System.currentTimeMillis()
+      (body, System.currentTimeMillis() - start)
+    }
+
+    def runAndFormat[A](id: SPProgramID, peer: Peer, f: (SPProgramID, Peer) => VcsAction[A])(showA: A => String): TryVcs[String] = {
+      val (out, time) = timed { f(id, peer).unsafeRun }
+      out.map { a => s"${showA(a)}\n$time ms"}
+    }
+
+    val add: VcsOp = (id, peer) =>
+      runAndFormat(id, peer, vcs.add) { _ => s"Added $id to $peer" }
+
+    val checkout: VcsOp = (id, peer) =>
+      runAndFormat(id, peer, vcs.checkout) { _ => s"Checked out $id from $peer" }
+
+    val pull: VcsOp = (id, peer) =>
+      runAndFormat(id, peer, vcs.pull) { updated =>
+        if (updated) "Updated local program." else "Already up to date."
+      }
+
+    val push: VcsOp = (id, peer) =>
+      runAndFormat(id, peer, vcs.push) { updated =>
+        if (updated) "Updated remote program." else "Already up to date."
+      }
+
+    val sync: VcsOp = (id, peer) =>
+      runAndFormat(id, peer, vcs.sync) {
+        case Neither    => "Already up to date."
+        case LocalOnly  => "Updated local program."
+        case RemoteOnly => "Updated remote program."
+        case Both       => "Synchronized."
+      }
+
+    val vcsOps: Map[String, VcsOp] = Map(
+      "add"      -> add,
+      "checkout" -> checkout,
+      "co"       -> checkout,
+      "pull"     -> pull,
+      "push"     -> push,
+      "sync"     -> sync
+    )
+
+    type VcsCmdHandler = (String, Array[String]) => Option[String]
+
+    val vcsCmdHandler: VcsCmdHandler = (cmd, args) =>
+      vcsOps.get(cmd).map { op =>
+        doVcsOp(cmd, args)(op)
+      }
+
+    val hostHandler: VcsCmdHandler = (cmd, args) =>
+      if ("host".equals(cmd))
+        args.length match {
+          case 0 => Some(host())
+          case 1 => Some(parseLoc(args(0)).map(host).toEither.merge)
+          case _ => None
+        }
+      else None
+
+    val handlers = List(vcsCmdHandler, hostHandler)
+
+    def vcs2(cmd: String, args: Array[String]): String =
+      try {
+        exec(cmd, args, handlers) | "Usage: vcs2 host|sync"
+      } catch {
+        case e: Exception =>
+          e.printStackTrace()
+          "???"
```

Comment:
Thanks, I need to clean that up a bit and add the other commands to the usage line.  Also, the "vcs2" command will be renamed to "vcs" once I get rid of the old version.